### ISSUE TITLE
IDE0380 removes required unsafe modifier from explicit-layout fields under updated memory-safety rules

### ISSUE THREAD

**Issue by EgorBo** [dotnet/roslyn#84564]
### Description

`IDE0380` incorrectly reports that `unsafe` is unnecessary on instance fields in an explicit-layout type when the updated memory-safety rules are enabled.

Under those rules, each such field must be explicitly marked `unsafe` or `safe`. Accepting the suggested **Remove unnecessary 'unsafe' modifier** code fix therefore makes the project fail with `CS9392`.

### Reproduction

Project settings:

```xml
<PropertyGroup>
  <OutputType>Exe</OutputType>
  <TargetFramework>net11.0</TargetFramework>
  <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
  <Features>$(Features);updated-memory-safety-rules</Features>
</PropertyGroup>
```

Source:

```csharp
using System.Runtime.InteropServices;

unsafe
{
    FooPair f = new();
    Test(ref f.f1, ref f.f2);
}

void Test(ref int g1, ref int g2) => g1 = g2;

[StructLayout(LayoutKind.Explicit)]
public struct FooPair
{
    [FieldOffset(0)] public unsafe int f1;
    [FieldOffset(1)] public unsafe int f2;
}
```

Interactive reproduction: [lab.razor.fyi](https://lab.razor.fyi/#jVLNbhMxEFYqVIFPqE8wSi-7KNk0uSA1SqU2TVCh0KhpQaKqhLM7ux3htVf-SbNFvAQnbog7N468CC_AFXgChHazbdMfUH2wx5-_z575POzbEmMjrRLN0yA0K1-WnCGZwDg3FtNg30lLKQY70qJW2Rj1lEI0XcacNDxGYO8YAMBQqREnDTH0QOKp53dL-ACN9TTGEAdxuwFV1PG77D1jU0XRJYOkhaTilHHHh94GJG3oQdLpMnY0ttqFdpfnyllvvjwjGQWDWSYoJOsfs8xNBIVgSuZ5UlWKR0NCEe3FsUHrrfnHUJGrQoo343b3JrV9O7VT1LByL-UkH_1YbrXg96cPn-FgZ7QOO9JY5BGoGOwJmQbkykHIJTiDEJHG0NIUDZCE_irEJNBUsMjBUxo2owg2YPuC6TcgVhpwxtNM4DprtQBgdT3TKkNtc-grGVPiNLekZG8bJy5hbA4G_fEJ19mIa4N7WXFuPDVfC3ursKw6eEX2ZJfLxPEEX6I2pKR3bR-MNE4JT_1LxUhjplWIxig9ztOJEsarbw-2Dp_UF1hD5NZpNN7c4HIuRtEsdWewqZURuWxa9RYlnaGuN6ButcO637hKHpLALW4wqrr2n0SXRdxi1EwxVTpvFl9n86Z2Ak2hqbI79pnfveZWX6UZidLNu3i2KYQ6PSx7wysyWSj7hROCTwT2lbQ4s-e33Q4HA1mAC_LiIKWzMpNdnKLwbiBB-d0LmnGGIcUUbhNPpDKWwvNXb7O-P24_XmvXG7CPmdL2UhSMXZZpNOa6raWicwfFVWcHKV2U_z8zC95ztDzilu9JkXsxFwbLq54uo9ZKm9cP7n__-fXPr-ThvTe1We1jrfYX)

### Actual behavior

The IDE offers `IDE0380`:

> Remove unnecessary 'unsafe' modifier

After accepting it, the declarations become:

```csharp
[FieldOffset(0)] public int f1;
[FieldOffset(1)] public int f2;
```

The compiler then reports:

```text
error CS9392: Field in an explicit or extended layout type must be marked 'unsafe' or 'safe'.
```

### Expected behavior

`IDE0380` should not be reported when removing `unsafe` would cause `CS9392`.

### Notes

The relevant implementation appears to be:

- `CSharpRemoveUnnecessaryUnsafeModifierDiagnosticAnalyzer`
- `UnnecessaryUnsafeModifierUtilities.ForkSemanticModelAndCheckNodes`

The analyzer removes candidate modifiers in a forked compilation and checks diagnostics within each modified declaration span. `CS9392` is validated from the containing type's field-layout checks, so it may not be observed by the current node-scoped diagnostic query.

Reproduced with .NET SDK `11.0.100-preview.6.26352.110`.

**Comment by CyrusNajmabadi** [comment]
>The analyzer removes candidate modifiers in a forked compilation and checks diagnostics within each modified declaration span. CS9392 is validated from the containing type's field-layout checks, so it may not be observed by the current node-scoped diagnostic query.

What's the location that the diagnostic is reported on?  Is it on the span of hte field itself?

**Comment by EgorBo** [comment]
> > The analyzer removes candidate modifiers in a forked compilation and checks diagnostics within each modified declaration span. CS9392 is validated from the containing type's field-layout checks, so it may not be observed by the current node-scoped diagnostic query.
> 
> What's the location that the diagnostic is reported on? Is it on the span of hte field itself?

Not sure I fully understand the question, isn't something that can be answered via the [labs](https://lab.razor.fyi/#jVLNbhMxEFaqCoFPqE8wSi-7KNk0uSA1SqU2TVCh0KhpQaKqhLM7ux3htVf-SbNFvAQnbog7N468CC_AFXgChHazbdMfUH2wx5-_z575POzbEmMjrRLN0yA0K1-WnCGZwDg3FtNg30lLKQY70qJW2Rj1lEI0XcacNDxGYO8YAMBQqREnDTH0QOKp53dL-ACN9TTGEAdxuwFV1PG77D1jU0XRJYOkhaTilHHHh94GJG3oQdLpMnY0ttqFdpfnyllvvjwjGQWDWSYoJOsfs8xNBIVgSuZ5UlWKR0NCEe3FsUHrrfnHUJGrQoo343b3JrV9O7VT1LCynHKSj37ca7Xg96cPn-FgZ7QOO9JY5BGoGOwJmQbkykHIJTiDEJHG0NIUDZCE_irEJNBUsMjBUxo2owg2YPuC6TcgVhpwxtNM4DprtQBgdT3TKkNtc-grGVPiNLekZG8bJy5hbA4G_fEJ19mIa4N7WXFuPDVfC3ursKw6eEX2ZJfLxPEEX6I2pKR3bR-MNE4JT_1LxUhjplWIxig9ztOJEsarbw-2Dp_UF1hD5NZpNN7c4HIuRtEsdWewqZURuWxa9RYlnaGuN6ButcO637hKHpLALW4wqrr2n0SXRdxi1EwxVTpvFl9n86Z2Ak2hqbI79pnfveZWX6UZidLNu3i2KYQ6PSx7wysyWSj7hROCTwT2lbQ4s-e33Q4HA1mAC_LiIKWzMpNdnKLwbiBB-d0LmnGGIcUUbhNPpDKWwvNXb7O-P24_XmvXG7CPmdL2UhSMXZZpNOa6raWicwfFVWcHKV2U_z8zC95ztDzilu9JkXsxFwbLq54uheb1g_vff3798yt5uPymNqt9rNX-Ag) link?

The issue here that Roslyn suggests an action that breaks compilation, once it's fixed it start suggesting to break it again.

**Comment by CyrusNajmabadi** [comment]
Ok.  So the error is reported on the field's identifier's span.  So this is an bug at the compiler layer.  The IDE asks specifically what diagnostics are reported in that location, and teh compiler does not report taht diagnostic.  

@RikkiGibson as i believe you were looking at this for other diagnostics as well.

**Comment by RikkiGibson** [comment]
I looked at this for nullable, I don't recall what my progress was there. I am assuming underlying cause is related in nature but the fix would be different.
